Sensors/SMCDellSensors/SMCDellSensors.hpp

```diff
@@ -0,0 +1,120 @@
+//
+//  SMCDellSensors.hpp
+//  SMCDellSensors
+//
+//  Copyright (C) 2001  Massimo Dal Zotto <dz@debian.org>
+//  http://www.debian.org/~dz/i8k/
+//  more work https://www.diefer.de/i8kfan/index.html , 2007
+//
+//  SMIMonitor ported from FakeSMC plugin (created by Slice 2014).
+//
+
+#ifndef SMCDellSensors_hpp
+#define SMCDellSensors_hpp
+
+#include "SMIMonitor.hpp"
+#include <Library/LegacyIOService.h>
+#include <IOKit/IOReportTypes.h>
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Winconsistent-missing-override"
+#include <IOKit/acpi/IOACPIPlatformDevice.h>
+#pragma clang diagnostic pop
+#include <Sensors/Private/pwr_mgt/IOPMPowerSource.h>
+#include <Sensors/Private/pwr_mgt/RootDomain.h>
+#include <IOKit/IOTimerEventSource.h>
+
+#include <Headers/kern_util.hpp>
+
+class EXPORT SMCDellSensors : public IOService {
+	OSDeclareDefaultStructors(SMCDellSensors)
+
+	/**
+	 *  Key name index mapping
+	 */
+	static constexpr size_t MaxIndexCount = sizeof("0123456789ABCDEF") - 1;
+	static constexpr const char *KeyIndexes = "0123456789ABCDEF";
+
+	/**
+	 *  Key name declarations
+	 */
+	
+	static constexpr SMC_KEY KeyFNum = SMC_MAKE_IDENTIFIER('F','N','u','m');
+	static constexpr SMC_KEY KeyF0ID(size_t i) { return SMC_MAKE_IDENTIFIER('F',KeyIndexes[i],'I','D'); }
+	static constexpr SMC_KEY KeyF0Ac(size_t i) { return SMC_MAKE_IDENTIFIER('F',KeyIndexes[i],'A','c'); }
+	//static constexpr SMC_KEY KeyF0As(size_t i) { return SMC_MAKE_IDENTIFIER('F',KeyIndexes[i],'A','s'); }
+	static constexpr SMC_KEY KeyF0Mn(size_t i) { return SMC_MAKE_IDENTIFIER('F',KeyIndexes[i],'M','n'); }
+	static constexpr SMC_KEY KeyF0Mx(size_t i) { return SMC_MAKE_IDENTIFIER('F',KeyIndexes[i],'M','x'); }
+	static constexpr SMC_KEY KeyF0Md(size_t i) { return SMC_MAKE_IDENTIFIER('F',KeyIndexes[i],'M','d'); }
+	static constexpr SMC_KEY KeyF0Tg(size_t i) { return SMC_MAKE_IDENTIFIER('F',KeyIndexes[i],'T','g'); }
+	static constexpr SMC_KEY KeyFFRC = SMC_MAKE_IDENTIFIER('F','S','!',' ');
```

Comment:
Why is `FS! ` called `FFRC`?

Reply:
Since exclamation mark cannot be used in class name. And this is FAN "force".

Reply:
Maybe rename it to something that doesn't look like key identifier, like `KeyFForce`? Another possible variant that hints the actual key name is `KeyFS__`.